src/metrics/rules/active_ruleset.go

```diff
@@ -522,11 +522,12 @@ func (as *activeRuleSet) matchRollupTarget(
 	}
 
 	var (
-		rollupTags    = rollupOp.Tags
-		sortedTagIter = matchOpts.SortedTagIteratorFn(sortedTagPairBytes)
-		matchTagIdx   = 0
-		nameTagName   = as.tagsFilterOpts.NameTagKey
-		nameTagValue  []byte
+		rollupTags      = rollupOp.Tags
+		sortedTagIter   = matchOpts.SortedTagIteratorFn(sortedTagPairBytes)
+		matchTagIdx     = 0
+		nameTagName     = as.tagsFilterOpts.NameTagKey
+		nameTagValue    []byte
+		includeTagNames = as.tagsFilterOpts.IncludeTagKeys
```

Comment:
How about let's introduce another variable to record the matchedIncludeTags?
- In the for loop when we have a match, we just include it in the matched variable
- in the very end of the function, we loop through the include tags and append them to the tagPairs

Reply:
One solution is to add the tagName tagVal pair to a map as we iterate through the for loop. Then at the end of the for loop, we iterate through the map and append to tag pairs as you mentioned. However, the problem is that iterating through a map doesn't guarantee order of iteration. This means that the resulting rollup id will not always have the tags in the same order each time we execute this function.

Other solutions are to create a custom struct to hold the tag pair information and create a slice to store these matched tag pairs, or to use an ordered map with log(N) key lookup.

Ultimately, I ended up creating a boolean flag that represents if we have a match on the include tags. If the flag is true then we update the tag pairs inside the for loop instead of outside the for loop. My most recent commit attempts to handle all the cases in the for loop:

1. If we have matched all the rollup rules and are still iterating through the tags in metric id, we still want to check for include tags in the remaining tags and add to tag pairs (which will be added to the rollup id).

2. If the current tag is a rollup tag and include tag, I don't modify the existing code. That tag will be added to the tagPairs as usual, and matchTagIdx will be updated accordingly.

3. If the current tag is a rollup tag and is **not** an include tag, we enter the same code path as case 2. I did not modify the code there, and this tag will still be added to tag pairs since it matches the rollup tag. matchTagIdx will be updated accordingly.

4. If the current tag is **not** a rollup tag and is an include tag, I wrote a check below the rollup tag matching code path that checks if the current tag is an include tag. We simply add the tag to the tag pairs because we want to add these include tags to the tag pairs. We continue so that we don't hit case 5 and return nil.

5. If the current tag is **not** a rollup tag and is **not** an include tag, we hit the end of the for loop and return nil because the current tag doesn't match anything.